pkg/listwatch/listwatch.go

```diff
@@ -44,56 +50,144 @@ import (
+			Resource: "namespaces",
+			Verbs:    []string{"list", "watch"},
+		},
+	)
+	if err != nil {
+		return nil, false, err
+	}
+
+	// The "kubernetes.io/metadata.name" label is GA since Kubernetes 1.22.
+	var metadataNameLabelSupported bool
+	v, err := semver.ParseTolerant(k8sVersion.String())
+	if err != nil {
+		level.Warn(l).Log("msg", "failed to parse Kubernetes version", "version", k8sVersion.String(), "err", err)
+	} else {
+		metadataNameLabelSupported = v.GTE(semver.MustParse("1.22.0"))
+	}
+
 	if IsAllNamespaces(allowedNamespaces) {
+		if !listWatchAllowed {
+			err := fmt.Errorf("missing list/watch permissions on the 'namespaces' resource")
+			for _, r := range reasons {
+				err = fmt.Errorf("%w: %w", err, r)
+			}
+
+			return nil, false, err
+		}
+
+		// deniedNamespaces is only supported with allowedNamespaces = "all".
 		tweak := func(options *metav1.ListOptions) {
 			DenyTweak(options, "metadata.name", deniedNamespaces)
 		}
+		if metadataNameLabelSupported {
+			// Using a label selector is more efficient but requires Kubernetes 1.22 at least.
+			tweak = func(options *metav1.ListOptions) {
+				TweakByLabel(options, "kubernetes.io/metadata.name", ExcludeFilterType, deniedNamespaces)
+			}
+		}
 
-		return cache.NewFilteredListWatchFromClient(c, "namespaces", metav1.NamespaceAll, tweak)
+		level.Debug(l).Log("msg", "using privileged namespace lister/watcher")
+		return cache.NewFilteredListWatchFromClient(
+			corev1Client.RESTClient(),
+			"namespaces",
+			metav1.NamespaceAll,
+			tweak,
+		), true, nil
+	}
+
+	if listWatchAllowed && metadataNameLabelSupported {
+		level.Debug(l).Log("msg", "using privileged namespace lister/watcher")
+		return cache.NewFilteredListWatchFromClient(
+			corev1Client.RESTClient(),
+			"namespaces",
+			metav1.NamespaceAll,
+			func(options *metav1.ListOptions) {
+				TweakByLabel(options, "kubernetes.io/metadata.name", IncludeFilterType, allowedNamespaces)
+			},
+		), true, nil
+	}
+
+	// At this point, the operator has no list/watch permissions on the
+	// namespaces resource. Check if it has at least the get permission to
+	// emulate the list/watch operations.
+	attrs := make([]k8sutil.ResourceAttribute, 0, len(allowedNamespaces))
+	for n := range allowedNamespaces {
+		attrs = append(attrs, k8sutil.ResourceAttribute{
+			Verbs:    []string{"get"},
+			Resource: "namespaces",
+			Name:     n,
+		})
+	}
+
+	getAllowed, reasons, err := k8sutil.IsAllowed(
+		ctx,
+		ssarClient,
+		nil, // namespaces is a cluster-scoped resource.
+		attrs...,
+	)
+	if err != nil {
+		return nil, false, err
+	}
+
+	// Only log a warning to preserve backward compatibility.
+	if !getAllowed {
+		err := fmt.Errorf("missing permissions")
+		for _, r := range reasons {
+			err = fmt.Errorf("%w: %w", err, r)
+		}
+
+		level.Warn(l).Log("msg", "the operator lacks required permissions which may result in degraded functionalities", "err", err)
 	}
 
 	listFunc := func(options metav1.ListOptions) (runtime.Object, error) {
```

Comment:
```suggestion
	listFunc := func(_ metav1.ListOptions) (runtime.Object, error) {
```

Is it intended to not use the options provided here?

Reply:
yes it's intended and the PR didn't change anything in that regard.